src/language-server/language-server.ts

```diff
@@ -0,0 +1,258 @@
+          };
+          if (a.type) {
+            item.detail = `{${a.type}}`;
+          }
+          if (a.inheritedFrom) {
+            if (item.detail) {
+              item.detail = `${item.detail} ⊃ ${a.inheritedFrom}`;
+            } else {
+              item.detail = `⊃ ${a.inheritedFrom}`;
+            }
+          }
+          return item;
+        }),
+      };
+    }
+    return {isIncomplete: false, items: []};
+  };
+
+  async getDefinition(textPosition: TextDocumentPositionParams):
+      Promise<Definition|undefined> {
+    const localPath =
+        this.converter.getWorkspacePathToFile(textPosition.textDocument);
+    const location =
+        await this._editorService.getDefinitionForFeatureAtPosition(
+            localPath, this.converter.convertPosition(textPosition.position));
+    if (location && location.file) {
+      let definition: Location = {
+        uri: this.converter.getUriForLocalPath(location.file),
+        range: this.converter.convertRange(location)
+      };
+      return definition;
+    }
+  }
+
+  async getDocsForHover(textPosition: TextDocumentPositionParams):
+      Promise<Hover|undefined> {
+    const localPath =
+        this.converter.getWorkspacePathToFile(textPosition.textDocument);
+    const documentation = await this._editorService.getDocumentationAtPosition(
+        localPath, this.converter.convertPosition(textPosition.position));
+    if (documentation) {
+      return {contents: documentation};
+    }
+  }
+
+  async handleChangedDocument(document: TextDocument) {
+    const localPath = this.converter.getWorkspacePathToFile(document);
+    await this._editorService.fileChanged(localPath, document.getText());
+    await this.reportErrors();
+  }
+
+  async handleErrors<Result, Fallback>(
+      promise: Promise<Result>,
+      fallbackValue: Fallback): Promise<Result|Fallback> {
+    try {
+      return await promise;
+    } catch (err) {
+      // Ignore WarningCarryingExceptions, they're expected, and made visible
+      //   to the user in a useful way. All other exceptions should be logged
+      //   if possible.
+      if (!(err instanceof WarningCarryingException)) {
+        this._connection.console.warn(err.stack || err.message || err);
+      }
+      return fallbackValue;
+    }
+  }
+
+  async reportErrors() {
+    // TODO(rictic): allow the user to set in a setting that they like warnings
+    //     at the package level, in which case we should call something like
+    //     getWarningsForPackage instead of going through each open document.
+    for (const document of this._documents.all()) {
+      const localPath = this.converter.getWorkspacePathToFile(document);
+      const warnings = await this._editorService.getWarningsForFile(localPath);
+      this._connection.sendDiagnostics({
+        diagnostics: warnings.map(
+            this.converter.convertWarningToDiagnostic, this.converter),
+        uri: document.uri
+      });
+    }
+  }
+
+  dispose(): void {
```

Comment:
So it would contribute an array of disposables to the class, and call dispose on each of them?

Hm, I ran into trouble with definition generation, I guess this is where I'd need to repeat the interface?